File: mkdocs/snippets/devbook/mosaics/change_mosaic_supply.mjs
import { PrivateKey } from 'symbol-sdk';
import {
	NemFacade,
	calculateTransactionFee,
	descriptors,
	models
} from 'symbol-sdk/nem';

const NODE_URL = process.env.NODE_URL ||
	'http://libertalia.nemtest.net:7890';
console.log('Using node', NODE_URL);

// Helper function to announce a transaction
async function announceTransaction(payload, label) {
	const announcePath = '/transaction/announce';
	console.log(`Announcing ${label} to ${announcePath}`);
	const response = await fetch(`${NODE_URL}${announcePath}`, {
		method: 'POST',
		headers: { 'Content-Type': 'application/json' },
		body: payload
	});
	if (!response.ok)
		throw new Error(`HTTP ${response.status}`);
	const result = await response.json();
	console.log('  Result:', result.message);
	return result.message;
}

// Helper function to fetch the current mosaic supply
async function fetchSupply(mosaic) {
	const supplyPath = `/mosaic/supply?mosaicId=${mosaic}`;
	const supplyResponse = await fetch(`${NODE_URL}${supplyPath}`);
	const supplyInfo = await supplyResponse.json();
	return supplyInfo.supply;
}

// Helper function to wait for transaction confirmation
async function waitForConfirmation(transactionHash, label) {
	const statusPath = `/transaction/get?hash=${transactionHash}`;
	console.log(`Waiting for ${label} confirmation from`, statusPath);
	let isConfirmed = false;
	for (let attempt = 1; 120 >= attempt; ++attempt) {
		const response = await fetch(`${NODE_URL}${statusPath}`);
		if (!response.ok) {
			console.log('  Transaction status: pending');
			await new Promise(resolve => { setTimeout(resolve, 1000); });
		} else {
			const confirmed = await response.json();
			console.log(`${label} confirmed in block`,
				confirmed.meta.height);
			isConfirmed = true;
			break;
		}
	}
	if (!isConfirmed)
		console.warn(`${label} confirmation took too long.`);
}
// [>step-1]
const SIGNER_PRIVATE_KEY = process.env.SIGNER_PRIVATE_KEY ||
	'0000000000000000000000000000000000000000000000000000000000000000';
const signerKeyPair = new NemFacade.KeyPair(
	new PrivateKey(SIGNER_PRIVATE_KEY));

const facade = new NemFacade('testnet');
const signerAddress = facade.network.publicKeyToAddress(
	signerKeyPair.publicKey);
console.log('Signer address:', signerAddress.toString());

const namespaceName = process.env.NAMESPACE || 'my_namespace';
const mosaicName = process.env.MOSAIC || 'token';
const mosaicId = `${namespaceName}:${mosaicName}`;
console.log('Mosaic ID:', mosaicId);
// [<step-1]
try {
	// --- INCREASING SUPPLY (MINTING) ---
	console.log('\n--- Increasing supply (minting) ---');
	// [>step-2]
	console.log('Supply before minting:', await fetchSupply(mosaicId));

	const increaseTx = facade.createTransactionFromTypedDescriptor(
		new descriptors.MosaicSupplyChangeTransactionV1Descriptor(
			new descriptors.MosaicIdDescriptor(
				new descriptors.NamespaceIdDescriptor(namespaceName),
				mosaicName),
			models.MosaicSupplyChangeAction.INCREASE,
			new models.Amount(500n)),
		signerKeyPair.publicKey,
		0n,
		2 * 60 * 60);
	increaseTx.fee = new models.Amount(
		calculateTransactionFee(increaseTx));

	const increaseSignature = facade.signTransaction(
		signerKeyPair, increaseTx);
	const increasePayload = facade.transactionFactory.static
		.attachSignature(increaseTx, increaseSignature);
	console.log('Built supply increase transaction:');
	console.dir(increaseTx.toJson(), { colors: true });
	const increaseResult = await announceTransaction(
		increasePayload, 'supply increase');
	if ('SUCCESS' === increaseResult) {
		await waitForConfirmation(
			facade.hashTransaction(increaseTx).toString(),
			'supply increase');
		console.log('Supply after minting:', await fetchSupply(mosaicId));
	} else {
		console.log('Supply increase rejected');
	}
	// [<step-2]
	// --- DECREASING SUPPLY (BURNING) ---
	console.log('\n--- Decreasing supply (burning) ---');
	// [>step-3]
	const decreaseTx = facade.createTransactionFromTypedDescriptor(
		new descriptors.MosaicSupplyChangeTransactionV1Descriptor(
			new descriptors.MosaicIdDescriptor(
				new descriptors.NamespaceIdDescriptor(namespaceName),
				mosaicName),
			models.MosaicSupplyChangeAction.DECREASE,
			new models.Amount(500n)),
		signerKeyPair.publicKey,
		0n,
		2 * 60 * 60);
	decreaseTx.fee = new models.Amount(
		calculateTransactionFee(decreaseTx));

	const decreaseSignature = facade.signTransaction(
		signerKeyPair, decreaseTx);
	const decreasePayload = facade.transactionFactory.static
		.attachSignature(decreaseTx, decreaseSignature);
	console.log('Built supply decrease transaction:');
	console.dir(decreaseTx.toJson(), { colors: true });
	const decreaseResult = await announceTransaction(
		decreasePayload, 'supply decrease');
	if ('SUCCESS' === decreaseResult) {
		await waitForConfirmation(
			facade.hashTransaction(decreaseTx).toString(),
			'supply decrease');
		console.log('Supply after burning:', await fetchSupply(mosaicId));
	} else {
		console.log('Supply decrease rejected');
	}
	// [<step-3]
} catch (e) {
	console.error(e.message, '| Cause:', e.cause?.code ?? 'unknown');
}

File: mkdocs/snippets/devbook/mosaics/ChangeMosaicSupply.java
//JAVA 21+
//DEPS org.symbol:symbol-sdk:3.3.3

import static java.nio.charset.StandardCharsets.UTF_8;

import java.io.IOException;
import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.net.http.HttpResponse.BodyHandlers;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;

import org.symbol.sdk.CryptoTypes;
import org.symbol.sdk.facade.NemFacade;
import org.symbol.sdk.nem.FeeCalculator;
import org.symbol.sdk.nem.KeyPair;
import org.symbol.sdk.nem.NemTransactionFactory;
import org.symbol.sdk.nem.descriptors.*;
import org.symbol.sdk.nem.models.Amount;
import org.symbol.sdk.nem.models.MosaicSupplyChangeAction;
import org.symbol.sdk.nem.models.Transaction;

public final class ChangeMosaicSupply {
	private static final ObjectMapper JSON_MAPPER = new ObjectMapper();

	private static final HttpClient HTTP_CLIENT =
		HttpClient.newHttpClient();

	private final String nodeUrl = System.getenv().getOrDefault(
		"NODE_URL", "http://libertalia.nemtest.net:7890");

	private final NemFacade facade = new NemFacade("testnet");

	private String announceTransaction(
		final String payload,
		final String label
	) throws IOException, InterruptedException {
		final String announcePath = "/transaction/announce";
		System.out.printf("Announcing %s to %s%n", label, announcePath);
		final HttpRequest request = HttpRequest.newBuilder(
			URI.create(nodeUrl + announcePath))
			.header("Content-Type", "application/json")
			.POST(HttpRequest.BodyPublishers.ofString(payload))
			.build();
		final HttpResponse<String> response = HTTP_CLIENT.send(
			request, BodyHandlers.ofString());
		if (2 != response.statusCode() / 100)
			throw new IOException("HTTP " + response.statusCode());
		final String result = JSON_MAPPER.readTree(response.body())
			.get("message").asText();
		System.out.printf("  Result: %s%n", result);
		return result;
	}

	private long fetchSupply(
		final String mosaicId
	) throws IOException, InterruptedException {
		final String path = String.format(
			"/mosaic/supply?mosaicId=%s", mosaicId);
		final HttpRequest request = HttpRequest.newBuilder(
			URI.create(nodeUrl + path)).GET().build();
		final HttpResponse<String> response = HTTP_CLIENT.send(
			request, BodyHandlers.ofString());
		return JSON_MAPPER.readTree(response.body())
			.get("supply").asLong();
	}

	private void waitForConfirmation(
		final String transactionHash,
		final String label
	) throws IOException, InterruptedException {
		final String statusPath = String.format(
			"/transaction/get?hash=%s", transactionHash);
		System.out.printf("Waiting for %s confirmation from %s%n",
			label, statusPath);
		boolean isConfirmed = false;
		for (int attempt = 1; 120 >= attempt; ++attempt) {
			final HttpRequest request = HttpRequest.newBuilder(
				URI.create(nodeUrl + statusPath)).GET().build();
			final HttpResponse<String> response = HTTP_CLIENT.send(
				request, BodyHandlers.ofString());
			if (2 != response.statusCode() / 100) {
				System.out.println("  Transaction status: pending");
				Thread.sleep(1000);
			} else {
				final JsonNode confirmed = JSON_MAPPER.readTree(
					response.body());
				System.out.printf("%s confirmed in block %s%n", label,
					confirmed.get("meta").get("height").asText());
				isConfirmed = true;
				break;
			}
		}
		if (!isConfirmed)
			System.out.printf("%s confirmation took too long.%n", label);
	}

	public static void main(final String[] args) {
		try {
			new ChangeMosaicSupply().run();
		} catch (final Exception ex) {
			System.out.println(null == ex.getMessage()
				? ex.toString()
				: ex.getMessage());
		}
	}

	private void run() throws IOException, InterruptedException {
		System.out.printf("Using node %s%n", nodeUrl);

		// [>step-1]
		final String privateKey = System.getenv().getOrDefault(
			"SIGNER_PRIVATE_KEY", "0".repeat(64));
		final KeyPair signerKeyPair = new KeyPair(
			new CryptoTypes.PrivateKey(privateKey));
		System.out.printf("Signer address: %s%n",
			facade.network.publicKeyToAddress(
				signerKeyPair.getPublicKey()));
		final String namespaceName = System.getenv().getOrDefault(
			"NAMESPACE", "my_namespace");
		final String mosaicName = System.getenv().getOrDefault(
			"MOSAIC", "token");
		final String mosaicId = namespaceName + ":" + mosaicName;
		System.out.printf("Mosaic ID: %s%n", mosaicId);
		// [<step-1]
		final MosaicIdDescriptor mosaicIdDescriptor =
			new MosaicIdDescriptor(
				new NamespaceIdDescriptor(namespaceName.getBytes(UTF_8)),
				mosaicName.getBytes(UTF_8));

		// --- INCREASING SUPPLY (MINTING) ---
		System.out.println("\n--- Increasing supply (minting) ---");
		// [>step-2]
		System.out.printf("Supply before minting: %d%n",
			fetchSupply(mosaicId));
		final Transaction increaseTx =
			facade.createTransactionFromTypedDescriptor(
				new MosaicSupplyChangeTransactionV1Descriptor(
					mosaicIdDescriptor,
					MosaicSupplyChangeAction.INCREASE,
					new Amount(500)),
				signerKeyPair.getPublicKey(), 0, 2 * 60 * 60);
		increaseTx.setFee(new Amount(
			FeeCalculator.calculateTransactionFee(increaseTx)));

		final CryptoTypes.Signature increaseSignature =
			facade.signTransaction(signerKeyPair, increaseTx);
		final String increasePayload =
			NemTransactionFactory.attachSignature(
				increaseTx, increaseSignature);
		System.out.println("Built supply increase transaction:");
		System.out.println(JSON_MAPPER.writerWithDefaultPrettyPrinter()
			.writeValueAsString(increaseTx.toJson()));
		final String increaseResult = announceTransaction(
			increasePayload, "supply increase");
		if ("SUCCESS".equals(increaseResult)) {
			waitForConfirmation(facade.hashTransaction(increaseTx)
				.toString(), "supply increase");
			System.out.printf("Supply after minting: %d%n",
				fetchSupply(mosaicId));
		} else {
			System.out.println("Supply increase rejected");
		}
		// [<step-2]
		// --- DECREASING SUPPLY (BURNING) ---
		System.out.println("\n--- Decreasing supply (burning) ---");
		// [>step-3]
		final Transaction decreaseTx =
			facade.createTransactionFromTypedDescriptor(
				new MosaicSupplyChangeTransactionV1Descriptor(
					mosaicIdDescriptor,
					MosaicSupplyChangeAction.DECREASE,
					new Amount(500)),
				signerKeyPair.getPublicKey(), 0, 2 * 60 * 60);
		decreaseTx.setFee(new Amount(
			FeeCalculator.calculateTransactionFee(decreaseTx)));

		final CryptoTypes.Signature decreaseSignature =
			facade.signTransaction(signerKeyPair, decreaseTx);
		final String decreasePayload =
			NemTransactionFactory.attachSignature(
				decreaseTx, decreaseSignature);
		System.out.println("Built supply decrease transaction:");
		System.out.println(JSON_MAPPER.writerWithDefaultPrettyPrinter()
			.writeValueAsString(decreaseTx.toJson()));
		final String decreaseResult = announceTransaction(
			decreasePayload, "supply decrease");
		if ("SUCCESS".equals(decreaseResult)) {
			waitForConfirmation(facade.hashTransaction(decreaseTx)
				.toString(), "supply decrease");
			System.out.printf("Supply after burning: %d%n",
				fetchSupply(mosaicId));
		} else {
			System.out.println("Supply decrease rejected");
		} // [<step-3]
	}
}
